### DISCUSSION TITLE
Support for the Haskell programming language

### DISCUSSION THREAD

**Discussion by wenkokke** [cursorless-dev/cursorless#2186] [Ideas]
This discussion is for support of the Haskell programming language (#2178).

The top-level comments below are RFCs asking for feedback on what the correct behaviour is for Haskell.
Please use the emoji reactions—🎉, 👀, 🚀—to vote.

**Comment by wenkokke** [comment]
```haskell
-- RFC 1: What should "arg" match?
fst :: (a, b) -> a
fst tup@(x, y) = x
--  ^^^^^^^^^^     <- 🎉 the whole pattern
--  ^^^            <- 👀 only the name of the whole argument, if given
--  ^^^  ^  ^      <- 🚀 all names in the pattern
```

**Comment by wenkokke** [comment]
```haskell
-- RFC 2: What should "branch" match?
foo = bar
  where
    bar = 1
--   🎉 `foo = bar` and `bar = 1`
--   👀 `foo = bar where bar = 1`
--   🚀 `foo = bar where bar = 1` and `bar = 1`
```

**Comment by wenkokke** [comment]
```haskell
-- RFC 3: What should "condition" match?
bap :: Int -> Int
bap | 1 == 1, 2 == 2 = x
--   🎉 `1 == 1` and `2 == 2`
--   👀 `1 == 1, 2 == 2`
```

**Reply by pokey** [reply]
Unfortunately chatgpt failed me on this one and claimed it's invalid syntax:

> In Haskell, each guard should be a separate boolean expression followed by an = and the result for that guard. Also, the comma is not the correct delimiter for separating guards; each guard should be on its own line or separated by a new |.

Could you possibly explain what the code is doing?

**Reply by wenkokke** [reply]
The comma combines guards, which in this case is simply boolean AND, but permits much more complicated stuff such as assignment.

**Reply by pokey** [reply]
got it, I'd be tempted to support both 🎉 and 👀, where for 👀 the domain is `1 == 1, 2 == 2 = x`

**Reply by wenkokke** [reply]
> got it, I'd be tempted to support both 🎉 and 👀, where for 👀 the domain is `1 == 1, 2 == 2 = x`

Both under "condition"?

**Reply by pokey** [reply]
yeah all of these are "condition"
